From ad44bacbc8476b7623a75fda0a41743f6c3b1b02 Mon Sep 17 00:00:00 2001
From: Sveinung Kvilhaugsvik <sveinung84@users.sourceforge.net>
Date: Mon, 29 May 2017 10:10:23 +0200
Subject: [PATCH] Give freeciv-web access to unit upgrade cost.

Make it possible to ask about unit upgrade cost in the unit_action_query
packet and get a reply in the unit_action_answer packet. The bundled clients
should be ported to make use of it if the upgrade cost starts relying on
factors the player can't see.

See hrm Feature #664985

---
 client/packhand.c | 13 +++++++++++++
 server/unithand.c | 22 ++++++++++++++++++++++
 2 files changed, 35 insertions(+)

diff --git a/client/packhand.c b/client/packhand.c
index 60138e8..7340c7a 100644
--- a/client/packhand.c
+++ b/client/packhand.c
@@ -4465,6 +4465,19 @@ void handle_unit_action_answer(int diplomat_id, int target_id, int cost,
       action_selection_next_in_focus(diplomat_id);
     }
     break;
+  case ACTION_UPGRADE_UNIT:
+    if (pcity && client.conn.playing
+        && is_human(client.conn.playing)) {
+      /* TODO: The bundled clients will have to start showing the upgrade
+       * price sent from the server before it can be allowed to rely on
+       * things the player can't see. (Example: it becomes legal to upgrade
+       * a unit in a foreign city.) */
+
+      /* Getting unit upgrade cost from the server is currently only used by
+       * Freeciv-web.  */
+      log_error("Received upgrade unit price but can't forward it.");
+    }
+    break;
   case ACTION_NONE:
     log_debug("Server didn't respond to query.");
     action_selection_no_longer_in_progress(diplomat_id);
diff --git a/server/unithand.c b/server/unithand.c
index fbd243c..80438bd 100644
--- a/server/unithand.c
+++ b/server/unithand.c
@@ -2161,6 +2161,28 @@ void handle_unit_action_query(struct connection *pc,
       return;
     }
     break;
+  case ACTION_UPGRADE_UNIT:
+    if (pcity
+        && is_action_enabled_unit_on_city(action_type,
+                                          pactor, pcity)) {
+      struct unit_type *tgt_utype;
+      int upgr_cost;
+
+      tgt_utype = can_upgrade_unittype(pplayer, unit_type_get(pactor));
+      upgr_cost = unit_upgrade_price(pplayer,
+                                      unit_type_get(pactor), tgt_utype);
+
+      dsend_packet_unit_action_answer(pc,
+                                      actor_id, target_id,
+                                      upgr_cost, action_type);
+    } else {
+      illegal_action(pplayer, pactor, action_type,
+                     pcity ? city_owner(pcity) : NULL,
+                     NULL, pcity, NULL, ACT_REQ_PLAYER);
+      unit_query_impossible(pc, actor_id, target_id);
+      return;
+    }
+    break;
   case ACTION_SPY_TARGETED_SABOTAGE_CITY:
     if (pcity
         && is_action_enabled_unit_on_city(action_type,
-- 
2.1.4

